python/pyarrow/_compute.pyx

```diff
@@ -416,7 +416,8 @@ cdef class MetaFunction(Function):
 
 cdef _pack_compute_args(object values, vector[CDatum]* out):
     for val in values:
-        if isinstance(val, (list, np.ndarray)):
+        if isinstance(val, (list, np.ndarray)) or \
+                hasattr(val, '__arrow_array__'):
```

Comment:
Agreed. I opened [ARROW-15202](https://issues.apache.org/jira/browse/ARROW-15202) separately because `pa.array` also doesn't check for `__array__`. So not supporting `__arrow_array__` here seemed like a bug to me, whereas supporting `__array__` in general seemed more like a feature request.